File: src/main/scala/ru/megains/old/graph/ShaderProgram.java
package ru.megains.old.graph;

import org.joml.Matrix3f;
import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.joml.Vector4f;
import org.lwjgl.BufferUtils;

import java.nio.FloatBuffer;
import java.util.HashMap;
import java.util.Map;

import static org.lwjgl.opengl.GL20.*;

public class ShaderProgram {

    private final int programId;

    private int vertexShaderId;

    private int fragmentShaderId;

    private final Map<String, UniformData> uniforms;



    public ShaderProgram() throws Exception {

        programId = glCreateProgram();
        if (programId == 0) {
            throw new Exception("Could not create Shader");
        }
        uniforms = new HashMap<>();
    }


    public void createUniform(String uniformName) throws Exception {

        int uniformLocation = glGetUniformLocation(programId, uniformName);
        if (uniformLocation < 0) {
            throw new Exception("Could not find uniform:" + uniformName);
        }
        uniforms.put(uniformName, new UniformData(uniformLocation));
    }


//    public void createFogUniform(String uniformName) throws Exception {
//        createUniform(uniformName + ".active");
//        createUniform(uniformName + ".colour");
//        createUniform(uniformName + ".density");
//    }

    public void setUniform(String uniformName, Matrix4f value) {
        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        // Check if float buffer has been created
        FloatBuffer fb = uniformData.getFloatBuffer();
        if (fb == null) {
            fb = BufferUtils.createFloatBuffer(16);
            uniformData.setFloatBuffer(fb);
        }
        // Dump the matrix into a float buffer
        value.get(fb);
        glUniformMatrix4fv(uniformData.getUniformLocation(), false, fb);
    }

    public void setUniform(String uniformName, Matrix3f value) {
        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        // Check if float buffer has been created
        FloatBuffer fb = uniformData.getFloatBuffer();
        if (fb == null) {
            fb = BufferUtils.createFloatBuffer(9);
            uniformData.setFloatBuffer(fb);
        }
        // Dump the matrix into a float buffer
        value.get(fb);
        glUniformMatrix3fv(uniformData.getUniformLocation(), false, fb);
    }




    public void setUniform(String uniformName, int value) {

        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
           // System.out.println("Uniform [" + uniformName + "] has nor been created");
            return;
          //  throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        glUniform1i(uniformData.getUniformLocation(), value);
    }

    public void setUniform(String uniformName, float value) {
        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        glUniform1f(uniformData.getUniformLocation(), value);
    }

    public void setUniform(String uniformName, boolean value) {


        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            // System.out.println("Uniform [" + uniformName + "] has nor been created");
            return;
            //  throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        glUniform1i(uniformData.getUniformLocation(), value ? 1 : 0);
    }

    public void setUniform(String uniformName, Vector3f value) {
        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        glUniform3f(uniformData.getUniformLocation(), value.x, value.y, value.z);
    }
    public void setUniform(String uniformName, Vector4f value) {
        UniformData uniformData = uniforms.get(uniformName);
        if (uniformData == null) {
            throw new RuntimeException("Uniform [" + uniformName + "] has nor been created");
        }
        glUniform4f(uniformData.getUniformLocation(), value.x, value.y, value.z, value.w);
    }
//    public void setUniform(String uniformName, Fog fog) {
//        setUniform(uniformName + ".active", fog.isActive() ? 1 : 0);
//        setUniform(uniformName + ".colour", fog.getColour() );
//        setUniform(uniformName + ".density", fog.getDensity());
//    }

    public void createVertexShader(String shaderCode) throws Exception {
        vertexShaderId = createShader(shaderCode, GL_VERTEX_SHADER);
    }

    public void createFragmentShader(String shaderCode) throws Exception {
        fragmentShaderId = createShader(shaderCode, GL_FRAGMENT_SHADER);
    }

    protected int createShader(String shaderCode, int shaderType) throws Exception {
        int shaderId = glCreateShader(shaderType);
        if (shaderId == 0) {
            throw new Exception("Error creating shader. Code: " + shaderId);
        }

        glShaderSource(shaderId, shaderCode);
        glCompileShader(shaderId);

        if (glGetShaderi(shaderId, GL_COMPILE_STATUS) == 0) {
            throw new Exception("Error compiling Shader code: " + glGetShaderInfoLog(shaderId, 1024));
        }

        glAttachShader(programId, shaderId);

        return shaderId;
    }

    public void link() throws Exception {
        glLinkProgram(programId);
        if (glGetProgrami(programId, GL_LINK_STATUS) == 0) {
            throw new Exception("Error linking Shader code: " + glGetShaderInfoLog(programId, 1024));
        }

        glValidateProgram(programId);
        if (glGetProgrami(programId, GL_VALIDATE_STATUS) == 0) {
            System.err.println("Warning validating Shader code: " + glGetShaderInfoLog(programId, 1024));
        }

    }

    public void bind() {
        glUseProgram(programId);
    }

    public void unbind() {
        glUseProgram(0);
    }

    public void cleanup() {
        unbind();
        if (programId != 0) {
            if (vertexShaderId != 0) {
                glDetachShader(programId, vertexShaderId);
            }
            if (fragmentShaderId != 0) {
                glDetachShader(programId, fragmentShaderId);
            }
            glDeleteProgram(programId);
        }
    }



}

File: src/main/scala/ru/megains/tartess/renderer/shader/ShaderProgram.scala
package ru.megains.tartess.renderer.shader

import org.joml.Matrix4f
import org.lwjgl.opengl.GL20._
import ru.megains.tartess.utils.Vec3f

import scala.collection.mutable

class ShaderProgram {

    val programId: Int = glCreateProgram
    if (programId == 0) throw new Exception("Could not create Shader")

    var vertexShaderId:Int = 0
    var fragmentShaderId:Int  = 0

    var uniforms:mutable.HashMap[String, UniformData] = new mutable.HashMap()


    def createShader(shaderCode: String, shaderType: Int): Int = {
        val shaderId = glCreateShader(shaderType)
        if (shaderId == 0) throw new Exception("Error creating shader. Code: " + shaderId)
        glShaderSource(shaderId, shaderCode)
        glCompileShader(shaderId)
        if (glGetShaderi(shaderId, GL_COMPILE_STATUS) == 0) throw new Exception("Error compiling Shader code: " + glGetShaderInfoLog(shaderId, 1024))
        glAttachShader(programId, shaderId)
        shaderId
    }

    def createVertexShader(shaderCode: String): Unit = {
        vertexShaderId = createShader(shaderCode, GL_VERTEX_SHADER)
    }

    def createFragmentShader(shaderCode: String): Unit = {
        fragmentShaderId = createShader(shaderCode, GL_FRAGMENT_SHADER)
    }

    def link(): Unit = {
        glLinkProgram(programId)
        if (glGetProgrami(programId, GL_LINK_STATUS) == 0) throw new Exception("Error linking Shader code: " + glGetShaderInfoLog(programId, 1024))
        glValidateProgram(programId)
        if (glGetProgrami(programId, GL_VALIDATE_STATUS) == 0) println("Warning validating Shader code: " + glGetShaderInfoLog(programId, 1024))
    }


    def createUniform(uniformName: String): Unit = {
        val uniformLocation = glGetUniformLocation(programId, uniformName)
        if (uniformLocation < 0) throw new Exception("Could not find uniform:" + uniformName)
        uniforms.put(uniformName, new UniformData(uniformLocation))
    }

    def setUniform(uniformName: String, value: Matrix4f): Unit = {
        val uniformData = getUniform(uniformName)
        value.get(uniformData.floatBuffer)
        glUniformMatrix4fv(uniformData.uniformLocation, false, uniformData.floatBuffer)
    }

    def setUniform(uniformName: String, value: Boolean): Unit = {
        val uniformData = getUniform(uniformName)
        glUniform1i(uniformData.uniformLocation, if (value) 1 else 0)
    }

    def setUniform(uniformName: String, value: Float): Unit = {
        val uniformData = getUniform(uniformName)
        glUniform1f(uniformData.uniformLocation, value)
    }

    def setUniform(uniformName: String, value: Vec3f): Unit = {
        val uniformData = getUniform(uniformName)
        glUniform3f(uniformData.uniformLocation, value.x, value.y, value.z)
    }


    def getUniform(uniformName: String): UniformData ={
        val uniformData = uniforms(uniformName)
        if (uniformData == null) throw new RuntimeException("Uniform [" + uniformName + "] has nor been created")
        uniformData
    }

    def bind(): Unit = glUseProgram(programId)

    def unbind(): Unit = glUseProgram(0)


}
